torchao/quantization/marlin_qqq/__init__.py

```diff
@@ -0,0 +1,373 @@
+
+        # Quantize
+        q_w = torch.round(w / s_channel).int()
+        q_w = torch.clamp(q_w, -max_q_val, max_q_val)
+        # Compute ref (dequantized)
+        w_ref = q_w.half() * s_channel
+
+        s_group = torch.tensor([], dtype=torch.half)
+        # div 2 ** (8 - self.bits)) to offset right shift in unpacking
+        s_channel /= 2 ** (8 - num_bits)
+        s_channel = s_channel.reshape(-1, size_n).contiguous().to(torch.float)
+
+    return (
+        w_ref.to(device=orig_device),
+        q_w.to(device=orig_device),
+        s_group.to(device=orig_device),
+        s_channel.to(device=orig_device),
+    )
+
+
+def pack_to_marlin_qqq(
+    q_w: torch.Tensor,
+    s_group: torch.Tensor,
+    s_channel: torch.Tensor,
+    num_bits: int,
+    group_size: int,
+) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+    """Pack the quantized weights and scales to the marlin format.
+
+    Args:
+        q_w (torch.Tensor): The quantized weight.
+        s_group (torch.Tensor): The per-group quantization scale.
+        s_channel (torch.Tensor): The per-channel quantization scale.
+        num_bits (int): The number of bits used for quantization.
+        group_size (int): The group size of quantization.
+    Returns:
+        Tuple[torch.Tensor, torch.Tensor, torch.Tensor]: The packed quantized weight in marlin format, the packed per-group scale in marlin format, and the packed per-channel scale in marlin format.
+    """
+    in_features, out_features = q_w.shape
+
+    assert num_bits == 4, "Marlin QQQ only supports 4-bit for now."
+
+    # Reformat to marlin
+    marlin_qqq_q_w = _to_marlin_weights(
+        q_w, in_features, out_features, num_bits, group_size
+    )
+    marlin_qqq_s_group, marlin_qqq_s_channel = _to_marlin_scales(
+        s_group, s_channel, in_features, out_features, num_bits, group_size
+    )
+
+    return marlin_qqq_q_w, marlin_qqq_s_group, marlin_qqq_s_channel
+
+
+def _to_marlin_weights(
+    q_w: torch.Tensor,
+    size_k: int,
+    size_n: int,
+    num_bits: int,
+    group_size: int,
+) -> torch.Tensor:
+    """Converts a quantized weight tensor to the marlin format.
+
+    Args:
+        q_w (torch.Tensor): The quantized weight.
+        size_k (int): The number of input features.
+        size_n (int): The number of output features.
+        num_bits (int): The number of bits used for quantization.
+        group_size (int): The group size of quantization.
+    Returns:
+        torch.Tensor: The packed quantized weight in marlin format.
+    """
+    if group_size == -1:
+        group_size = size_k
+    quant_type = "per-channel" if group_size == size_k else "per-group"
+    # Permute
+    perm = get_qqq_weight_perm(num_bits, quant_type)
+    q_w = marlin_permute_weights(q_w, size_k, size_n, perm)
+
+    # Pack
+    pack_factor = get_pack_factor(num_bits)
+    orig_device = q_w.device
+
+    q_w = q_w.cpu().numpy().astype(numpy.uint32)
+
+    q_packed = numpy.zeros(
```

Comment:
curious why we go to numpy for this?

Reply:
Originally, `q_w` is `torch.uint32`, but since torch does not support `lshift_cuda` or `lshift_cpu`, we used `numpy+uint32` to implement the packing function. However, I have discovered that we can convert `q_w` to `torch.int64`, which supports `lshift_cuda`, and perform the packing on the GPU. This should be faster. I will update the code accordingly.

Reply:
I have updated the code.